### PR TITLE
Noexcept move constructors/assigment operators 

### LINKED ISSUE #2240: All move constructors and assignment operators in Kokkos should be `noexcept`
Almost all of Kokkos should be `noexcept`, but let's break it up into pieces that can be tackled in reasonable-sized issues.  This one is the most urgent.

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #2240.

### CODE REVIEW

**Comment by nliber** [comment]
The changes look fine.

Note:  defaulted constructors/assignment operators need not be marked `noexcept` as the language will do it automatically (assuming all the non-static member variables and base class corresponding functions are also `noexcept`).  I presume these are being marked explicitly because of clang-tidy.

**Comment by dalg24** [comment]
@masterleinad please confirm that clang-tidy complains when defaulted move constructors and assignment operators are not marked.

**Comment by masterleinad** [comment]
> @masterleinad please confirm that clang-tidy complains when defaulted move constructors and assignment operators are not marked.

Yes:
```
/tmp/kokkos_new/core/src/Kokkos_Crs.hpp:112:29: error: move constructors should be marked noexcept [performance-noexcept-move-constructor,-warnings-as-errors]
  KOKKOS_DEFAULTED_FUNCTION Crs(Crs&&) = default;
                            ^
                                        noexcept 
```
I remember some compilers complaining if the `noexcept`-specifier doesn't match with the implicitly generated one so I have a slight preference just adding it (as suggested by `clang-tidy`).

**Comment by masterleinad** [comment]
@nliber Of course, we should not mark anything as `noexcept` that actually throws. All I was saying is that some compilers warn if they deduce that the `noexcept` specifier doesn't match the implicitly generated one (as can be seen in https://travis-ci.org/github/kokkos/kokkos/builds/721138461?utm_source=github_status&utm_medium=notification). 

I am still in favor of being more expressive also for defaulted member functions following the clang-tidy check here for that reason. In particular, I think I marked at least all of the move assignment operators and move construction operators here (admittedly I didn't check in detail if all of them were correct but at least `clang` isn't complaining anymore).
If the overall consensus is that we should rather not do anything about defaulted member functions (and hence not running into the danger of `throw` implying `terminate`), I am happy to remove it, of course, and the `clang-tidy` check with it.

> noexcept has semantics: it means that if an exception tries to escape the function, std::terminate is called.
Isn't that rather the result violing the promise to not throw?

I looked at the issue in the first place since it was marked for the 3.2 release (or rather the next release) and think that the `clang-tidy` check is helpful. If the agreement is that this is not worth the effort at the moment, we can postpone it of course, and close this pull request instead.

**Comment by nliber** [comment]
Looking at one of the build errors:

> /Users/travis/build/kokkos/kokkos/core/src/impl/Kokkos_ViewMapping.hpp:3172:29: error: exception specification of explicitly defaulted move constructor does not match the calculated one
> 5959  KOKKOS_DEFAULTED_FUNCTION ViewMapping(ViewMapping&&) noexcept = default;

This error is caused because the `ViewMapping` move constructor is calculated to be `noexcept(false)` and we are explicitly marking it `noexcept` (aka `noexcept(true)`), isn't it?  If we removed the explicit `noexcept`, does the warning go away?

Diving into that:  `ViewMapping` has two non-static data members:  a `ViewOffset` and a `ViewDataHandle`.

Speculation (since I don't know the exact types being instantiated here):  some of the `ViewOffset` specializations have a non-defaulted copy constructor which is not marked `noexcept`, so it is calculated to be `noexcept(false)`.  The explicitly declared `ViewOffset` copy constructor also means that `ViewOffset` doesn't have a move constructor, so the defaulted `ViewMapping` move constructor will call the `ViewOffset` copy constructor which causes the mismatch.  The correct fix would be to mark these non-defaulted `ViewOffset` copy constructors as `noexcept`.

I'm a neophyte when it comes to clang-tidy.  Would it be possible to use it to find these problematic non-defaulted functions?

I'm all for removing the explicit `noexcept` on defaulted functions and adding explicit `noexcept` on the non-defaulted move/copy constructors/assignment operators and default constructors which don't throw.

**Comment by masterleinad** [comment]
Regarding the `ViewMapping` error: That's fixed by [`23a52cd` (#3306)](https://github.com/kokkos/kokkos/pull/3306/commits/23a52cdc915d7281332adef4cf4f13be7121bfe2) but the rule of 0 also works there.

**Comment by crtrott** [comment]
So does this mean that the compiler now has to bracket every one of these calls with a catch, in order to call std::terminate if it did throw?

**Comment by masterleinad** [comment]
> So does this mean that the compiler now has to bracket every one of these calls with a catch, in order to call std::terminate if it did throw?

A `noexcept` function will call `std::terminate` if it throws even if it is defaulted and the implicitly calculated `noexcept` specifier would be `noexcept(true)` (if that is what you are asking). 
`std::terminate` is also called when no catch handler can be found for a thrown exception (and `std::terminate` calls `abort` by default).

**Review by crtrott — CHANGES_REQUESTED** [review]
We need to do a bit more performance stuff. 

like:
```
void bar();

void foo() noexcept {
  bar();
}
```
inserts a catch and conditional terminate into foo. Is this innocent or does this cost us in practice more performance than we gain in situations where some standard libraries are called with Kokkos objects. 

```
foo():                                # @foo()
        push    rax
        call    bar()
        mov     rdi, rax
        call    __clang_call_terminate
__clang_call_terminate:                 # @__clang_call_terminate
        push    rax
        call    __cxa_begin_catch
        call    std::terminate()
```
https://godbolt.org/z/acbr6Y

**Comment by crtrott** [comment]
this is intel:
```
foo():
        push      rsi                                           #3.20
        call      bar()                                       #4.3
        pop       rcx                                           #5.1
        ret                                                     #5.1
        mov       QWORD PTR [rsp], rax                          #3.20
        xor       eax, eax                                      #3.20
        mov       rdi, QWORD PTR [rsp]                          #3.20
        call      __cxa_call_unexpected                         #3.20
        mov       rdi, QWORD PTR [rsp]                          #3.20
        call      _Unwind_Resume 
```

**Comment by DavidPoliakoff** [comment]
I think if we're doing a performance study, step 1 should be determining how much time we spend in functions we're considering marking noexcept. I don't want to spend a week figuring out how to optimize 0.1% of our users' runtime

**Comment by calewis** [comment]
@nliber, can you tell me what you want clang-tidy to do.  I partly followed this discussion.  But A statement of I want clang-tidy to detect X would help me.  My initial (short) attempt to understand what you want is to see if adding a non-defaulted function caused the defaulted functions to flip from `noexcept(true)` to `noexcept(false)` is that correct?

File: containers/src/Kokkos_ScatterView.hpp [hunk 1]
```diff
@@ -229,7 +229,7 @@ struct ScatterValue<ValueType, Kokkos::Experimental::ScatterSum, DeviceType,
  public:
   KOKKOS_FORCEINLINE_FUNCTION ScatterValue(ValueType& value_in)
       : value(value_in) {}
-  KOKKOS_FORCEINLINE_FUNCTION ScatterValue(ScatterValue&& other)
+  KOKKOS_FORCEINLINE_FUNCTION ScatterValue(ScatterValue&& other) noexcept
```
Comment (crtrott) [inline containers/src/Kokkos_ScatterView.hpp:232]: Depending on the compiler this now may insert a catch on every access to a scatterview which is performance critical code ...

File: core/src/Cuda/Kokkos_Cuda_View.hpp [hunk 2]
```diff
@@ -160,7 +160,7 @@ struct CudaLDGFetch {
   CudaLDGFetch(const CudaLDGFetch& rhs) : m_ptr(rhs.m_ptr) {}
 
   KOKKOS_INLINE_FUNCTION
-  CudaLDGFetch(CudaLDGFetch&& rhs) : m_ptr(rhs.m_ptr) {}
+  CudaLDGFetch(CudaLDGFetch&& rhs) noexcept : m_ptr(rhs.m_ptr) {}
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_View.hpp:163]: depending on compiler this now injects catch clauses into performance crticial code.

**Review by crtrott — CHANGES_REQUESTED** [review]
Adding noexcept to the lowest most access function for some stuff feels dangerous for the above mentioned reasons.

**Pull request closed without merging**